dcm4che-core/src/main/java/org/dcm4che3/io/DicomInputStream.java

```diff
@@ -861,10 +957,15 @@ private void readFragments(Attributes attrs, int fragsTag, VR vr)
     public byte[] readValue() throws IOException {
         int valLen = length;
         try {
-            if (valLen < 0)
-            throw new IOException(
-                "internal error: length should have been validated in readHeader");
-            int allocLen = allocateLimit >= 0
+            if (valLen < 0) {
+                throw new IOException("internal error: length should have been validated in readHeader");
+            }
+            boolean limitedStream = in instanceof LimitedInputStream;
+            if(limitedStream && valLen > ((LimitedInputStream)in).getRemaining()) {
+                throw new EOFException(String.format("Length %d for tag %s @ %d exceeds remaining %d (pos: %d)",
+                        valLen, TagUtils.toString(tag), tagPos, ((LimitedInputStream) in).getRemaining(), pos));
+            }
+            int allocLen = allocateLimit > 0 && !limitedStream
```

Comment:
I know, but I changed it to `allocateLimit != -1` in the latest commits, because now the invalid value of 0 gets already rejected in the setAllocateLimit() setter.